### PR TITLE
added template to flat index

### PR DESCRIPTION
**Pull request by meiravgri**
## changes in the pull request
This pr is part of adding fp64 support to vector similarity indexes.

### Templating flat index
**template <typename DataType, typename DistFuncType>**
Instead of hard coded float, Flat index is now templated with two types: 
- **DataType** vector data type
- **DistFuncType** is the type of the returned value from the distance calculations. 
this type depends on the _metric_ and the vector _data type._

### brute force index files arrangement
All brute force index functions implementations were **moved to the header file**, and the cpp file was deleted.

### includes
Since there is a bilateral dependency between BF batch iterator and BF index, **includes** in each file should be done **carefully**:
- **bf_batch_iterator.h:** incomplete declaration of BruteForceIndex
-  **bf_batch_iterator.cpp:** includes brute_force.h
- **brute_force.h:** includes bf_batch_iterator.h after the class definition and before the member function definitions.

## Main objects this PR modified
1. brute force index
2. VecSim_InfoField struct

**Mark if applicable**

- [x] This PR introduces API changes:
VecSim_InfoFieldType is holding a named union, meaning we should add fieldValue. before accessing the union fields.
- [ ] This PR introduces serialization changes

### CODE REVIEW

File: src/VecSim/algorithms/brute_force/brute_force.h [hunk 1]
```diff
@@ -83,3 +84,457 @@ class BruteForceIndex : public VecSimIndex {
...
+
+    VecSimQueryResult_List rl = {0};
+    void *timeoutCtx = queryParams ? queryParams->timeoutCtx : NULL;
+
+    this->last_mode = STANDARD_KNN;
+    DataType normalized_blob[this->dim]; // This will be use only if metric == VecSimMetric_Cosine.
+    if (this->metric == VecSimMetric_Cosine) {
+        // TODO: need more generic
+        memcpy(normalized_blob, queryBlob, this->dim * sizeof(DataType));
+        float_vector_normalize(normalized_blob, this->dim);
```
Comment (alonre24) [inline src/VecSim/algorithms/brute_force/brute_force.h:318]: Same here (is making it generic should be part of this PR? + avoid allocating on stack if not have to)

File: src/VecSim/algorithms/brute_force/brute_force.h [hunk 2]
```diff
@@ -83,3 +84,457 @@ class BruteForceIndex : public VecSimIndex {
...
+                                                    VecSimQueryParams *queryParams) {
+    auto rl = (VecSimQueryResult_List){0};
+    void *timeoutCtx = queryParams ? queryParams->timeoutCtx : nullptr;
+    this->last_mode = RANGE_QUERY;
+
+    DataType normalized_blob[this->dim]; // This will be use only if metric == VecSimMetric_Cosine.
+    if (this->metric == VecSimMetric_Cosine) {
+        // TODO: need more generic when other types will be supported.
+        memcpy(normalized_blob, queryBlob, this->dim * sizeof(DataType));
+        float_vector_normalize(normalized_blob, this->dim);
```
Comment (alonre24) [inline src/VecSim/algorithms/brute_force/brute_force.h:370]: Same here (is making it generic should be part of this PR? + avoid allocating on stack if not have to)

File: src/VecSim/algorithms/brute_force/brute_force.h [hunk 3]
```diff
@@ -83,3 +84,457 @@ class BruteForceIndex : public VecSimIndex {
...
+}
+
+template <typename DataType, typename DistFuncType>
+VecSimBatchIterator *
+BruteForceIndex<DataType, DistFuncType>::newBatchIterator(const void *queryBlob,
+                                                          VecSimQueryParams *queryParams) {
+    auto *queryBlobCopy = this->allocator->allocate(sizeof(DataType) * this->dim);
+    memcpy(queryBlobCopy, queryBlob, dim * sizeof(DataType));
+    if (metric == VecSimMetric_Cosine) {
+        float_vector_normalize((DataType *)queryBlobCopy, dim);
```
Comment (alonre24) [inline src/VecSim/algorithms/brute_force/brute_force.h:462]: Same here (is making it generic should be part of this PR?)

**Review by alonre24 — COMMENTED** [review]
Looks very good and clear!
Added some comments and suggestions

File: src/VecSim/algorithms/brute_force/brute_force.h [hunk 4]
```diff
@@ -83,3 +84,457 @@ class BruteForceIndex : public VecSimIndex {
...
+}
+
+template <typename DataType, typename DistFuncType>
+int BruteForceIndex<DataType, DistFuncType>::addVector(const void *vector_data, size_t label) {
+
+    DataType normalized_data[this->dim]; // This will be use only if metric == VecSimMetric_Cosine
+    if (this->metric == VecSimMetric_Cosine) {
+        // TODO: need more generic
+        memcpy(normalized_data, vector_data, this->dim * sizeof(DataType));
+        float_vector_normalize(normalized_data, this->dim);
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/algorithms/brute_force/brute_force.h:152]: No, ill do it in a different PR

File: src/VecSim/algorithms/brute_force/brute_force.h [hunk 5]
```diff
@@ -83,3 +84,457 @@ class BruteForceIndex : public VecSimIndex {
...
+    size_t index = getVectorRelativeIndex(id);
+
+    // Update vector data in the block.
+    vectorBlock->updateVector(index, vector_data);
+}
+
+template <typename DataType, typename DistFuncType>
+int BruteForceIndex<DataType, DistFuncType>::addVector(const void *vector_data, size_t label) {
+
+    DataType normalized_data[this->dim]; // This will be use only if metric == VecSimMetric_Cosine
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/algorithms/brute_force/brute_force.h:148]: cool thanks for reminding me
ill do it in the implementation of the templated normalize function

File: src/VecSim/algorithms/brute_force/bf_batch_iterator.cpp [hunk 6]
```diff
@@ -7,6 +7,7 @@
 #include "VecSim/utils/arr_cpp.h"
 #include "VecSim/utils/vec_utils.h"
 #include "VecSim/query_result_struct.h"
+#include "VecSim/algorithms/brute_force/brute_force.h"
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/algorithms/brute_force/bf_batch_iterator.cpp:10]: see PR description

File: src/VecSim/algorithms/brute_force/bf_batch_iterator.h [hunk 7]
```diff
@@ -1,14 +1,18 @@
-#include "VecSim/batch_iterator.h"
-#include "VecSim/algorithms/brute_force/brute_force.h"
+#pragma once
 
+#include "VecSim/batch_iterator.h"
+#include "VecSim/utils/vecsim_stl.h"
+#include "vector_block.h"
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/algorithms/brute_force/bf_batch_iterator.h:5]: se PR description

File: src/VecSim/algorithms/brute_force/brute_force.h [hunk 8]
```diff
@@ -45,7 +46,7 @@ class BruteForceIndex : public VecSimIndex {
     } // throws out_of_range
 
     inline vecsim_stl::vector<VectorBlock *> getVectorBlocks() const { return vectorBlocks; }
-    inline dist_func_t<float> distFunc() const { return dist_func; }
+    inline dist_func_t<DistFuncType> distFunc() const { return dist_func; }
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/algorithms/brute_force/brute_force.h:49]: removed this function

**Review by alonre24 — APPROVED** [review]

File: src/VecSim/vec_sim.cpp [hunk 9]
```diff
@@ -11,7 +11,7 @@
 #include "memory.h"
 
 extern "C" void VecSim_SetTimeoutCallbackFunction(timeoutCallbackFunction callback) {
-    VecSimIndexAbstract::setTimeoutCallbackFunction(callback);
+    VecSimIndexAbstract<float>::setTimeoutCallbackFunction(callback);
 }
```
Comment (GuyAv46) [inline src/VecSim/vec_sim.cpp:15]: This function is called once from RediSearch. should it handle all template types? perhaps we should move the static callback to the interface and avoid the template here?

File: src/VecSim/algorithms/brute_force/brute_force.h [hunk 10]
```diff
@@ -73,3 +76,400 @@ class BruteForceIndex : public VecSimIndexAbstract {
...
+/******************************* Implementation **********************************/
+
+#include "VecSim/spaces/spaces.h"
+#include "VecSim/query_result_struct.h"
+#include "VecSim/algorithms/brute_force/bf_batch_iterator.h"
+
+#include <cstring>
+#include <cmath>
+
+using namespace std;
```
Comment (DvirDukhan) [inline src/VecSim/algorithms/brute_force/brute_force.h:89]: move to the head of the file
also avoid `using namespace std` but have `using std::classname`

File: src/VecSim/algorithms/brute_force/brute_force.h [hunk 11]
```diff
@@ -73,3 +76,400 @@ class BruteForceIndex : public VecSimIndexAbstract {
...
+                             .stringValue = VecSimSearchMode_ToString(info.bfInfo.last_mode)}}});
+
+    return infoIterator;
+}
+
+template <typename DataType, typename DistType>
+VecSimBatchIterator *
+BruteForceIndex<DataType, DistType>::newBatchIterator(const void *queryBlob,
+                                                      VecSimQueryParams *queryParams) {
+    // As this is the only supported type, we always allocate 4 bytes for every element in the
```
Comment (DvirDukhan) [inline src/VecSim/algorithms/brute_force/brute_force.h:391]: remove this comment?

File: src/VecSim/algorithms/brute_force/brute_force_single.h [hunk 12]
```diff
@@ -51,3 +53,70 @@ class BruteForceIndex_Single : public BruteForceIndex {
     friend class BM_VecSimBasics_DeleteVectorBF_Benchmark;
 #endif
 };
+
+/******************************* Implementation **********************************/
+
+#include "VecSim/utils/vec_utils.h"
+#include "VecSim/query_result_struct.h"
```
Comment (DvirDukhan) [inline src/VecSim/algorithms/brute_force/brute_force_single.h:60]: move to the head of the file

File: src/VecSim/info_iterator.h [hunk 13]
```diff
@@ -27,12 +34,7 @@ typedef enum {
 typedef struct {
     const char *fieldName;          // Field name.
     VecSim_InfoFieldType fieldType; // Field type (in {STR, INT64, FLOAT64})
-    union {
-        double floatingPointValue; // Floating point value. 64 bits float.
-        int64_t integerValue;      // Integer value. Signed 64 bits integer.
-        u_int64_t uintegerValue;   // Unsigned value. Unsigned 64 buts integer.
-        const char *stringValue;   // String value.
-    };
+    FieldValue fieldValue;
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/info_iterator.h:37]: mm yes

File: src/VecSim/algorithms/brute_force/bf_batch_iterator.h [hunk 14]
```diff
@@ -1,14 +1,18 @@
...
 #include <limits>
 
 using namespace std;
 
+template <typename DataType, typename DistType>
+class BruteForceIndex;
 class BF_BatchIterator : public VecSimBatchIterator {
 private:
-    const BruteForceIndex *index;
+    const BruteForceIndex<float, float> *index;
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/algorithms/brute_force/bf_batch_iterator.h:15]: in the next PR

File: src/VecSim/info_iterator.h (hunk 13 above)
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/info_iterator.h:37]: we can either remove the union and always keep all the fields, 
or we can update this change in Redisearch (FT.DEBUG uses this struct)

**Review by DvirDukhan — APPROVED** [review]

**Pull request merged**